File: src/providers/firecrawl.ts
/**
 * Firecrawl — clean a URL (scrape) or crawl a site.
 *
 * v0.2 implements scrape end-to-end, plus startCrawl/pollCrawl/crawl with a
 * bounded poll loop. ops/crawl.ts wraps it with the apply gates.
 */

import {
  FetchedPage,
  MissingKeyError,
  NormalizedResult,
  Provider,
  ProviderError,
  httpRequest,
  safeGet,
} from "./base.js";

export const BASE_URL = "https://api.firecrawl.dev/v1";

export interface ScrapeOptions {
  formats?: string[];
  onlyMainContent?: boolean;
  screenshot?: boolean;
  timeoutMs?: number;
}

export interface CrawlOptions {
  limit?: number;
  includePaths?: string[];
  excludePaths?: string[];
  formats?: string[];
  pollIntervalMs?: number;
  maxWaitMs?: number;
}

export class FirecrawlProvider extends Provider {
  name = "firecrawl";
  schemaVersion = "firecrawl-v1-2026-04";
  apiKey: string | undefined;

  constructor(apiKey?: string) {
    super();
    this.apiKey = apiKey;
  }

  private ensureKey(): string {
    if (!this.apiKey) throw new MissingKeyError(this.name);
    return this.apiKey;
  }

  private authHeaders(): Record<string, string> {
    return { authorization: `Bearer ${this.ensureKey()}` };
  }

  async scrape(url: string, opts: ScrapeOptions = {}): Promise<FetchedPage> {
    const formats = opts.formats?.length ? [...opts.formats] : ["markdown"];
    if (opts.screenshot && !formats.includes("screenshot")) formats.push("screenshot");
    const body: Record<string, unknown> = {
      url,
      formats,
      onlyMainContent: opts.onlyMainContent ?? true,
    };
    const response = await httpRequest(`${BASE_URL}/scrape`, {
      method: "POST",
      headers: this.authHeaders(),
      body,
      timeoutMs: opts.timeoutMs ?? 60_000,
    });
    return this.normalizeScrape((response.json as Record<string, unknown>) ?? {}, url);
  }

  private normalizeScrape(payload: Record<string, unknown>, requestedUrl: string): FetchedPage {
    if (!safeGet<boolean>(payload, ["success"], false)) {
      const err = safeGet<string>(payload, ["error"]) ?? "scrape failed";
      throw new ProviderError(`firecrawl: ${err}`);
    }
    const data = (safeGet<Record<string, unknown>>(payload, ["data"], {}) ?? {}) as Record<string, unknown>;
    const url = (safeGet<string>(data, ["metadata", "sourceURL"]) ?? requestedUrl) || requestedUrl;
    return new FetchedPage({
      url,
      title: safeGet<string>(data, ["metadata", "title"], "") ?? "",
      markdown: safeGet<string>(data, ["markdown"], "") ?? "",
      html: safeGet<string>(data, ["html"]) ?? null,
      metadata: (safeGet<Record<string, unknown>>(data, ["metadata"], {}) ?? {}) as Record<string, unknown>,
      provider: this.name,
      fetchedAt: safeGet<string>(data, ["metadata", "fetchTime"]) ?? null,
      status: "ok",
    });
  }

  async startCrawl(url: string, opts: CrawlOptions = {}): Promise<string> {
    const body: Record<string, unknown> = {
      url,
      limit: opts.limit ?? 10,
      scrapeOptions: { formats: opts.formats ?? ["markdown"] },
    };
    if (opts.includePaths) body.includePaths = opts.includePaths;
    if (opts.excludePaths) body.excludePaths = opts.excludePaths;
    const response = await httpRequest(`${BASE_URL}/crawl`, {
      method: "POST",
      headers: this.authHeaders(),
      body,
      timeoutMs: 60_000,
    });
    const payload = (response.json as Record<string, unknown>) ?? {};
    if (!safeGet<boolean>(payload, ["success"], false)) {
      throw new ProviderError(`firecrawl: crawl failed to start: ${JSON.stringify(safeGet(payload, ["error"]))}`);
    }
    const id = safeGet<string>(payload, ["id"]);
    if (!id) throw new ProviderError(`firecrawl: crawl response missing id`);
    return id;
  }

  async pollCrawl(jobId: string): Promise<Record<string, unknown>> {
    const response = await httpRequest(`${BASE_URL}/crawl/${jobId}`, {
      method: "GET",
      headers: this.authHeaders(),
      timeoutMs: 30_000,
    });
    return ((response.json as Record<string, unknown>) ?? {}) as Record<string, unknown>;
  }

  async crawl(url: string, opts: CrawlOptions = {}): Promise<FetchedPage[]> {
    const id = await this.startCrawl(url, opts);
    const interval = opts.pollIntervalMs ?? 2000;
    const deadline = Date.now() + (opts.maxWaitMs ?? 300_000);
    while (Date.now() < deadline) {
      const status = await this.pollCrawl(id);
      const state = safeGet<string>(status, ["status"]);
      if (state === "completed" || state === "failed") {
        if (state === "failed") {
          throw new ProviderError(`firecrawl: crawl ${id} failed: ${JSON.stringify(safeGet(status, ["error"]))}`);
        }
        const items = (safeGet<unknown[]>(status, ["data"], []) ?? []) as Array<Record<string, unknown>>;
        return items.map((item) => this.pageFromCrawlItem(item, id));
      }
      await sleep(interval);
    }
    throw new ProviderError(`firecrawl: crawl ${id} did not complete in ${opts.maxWaitMs ?? 300_000}ms`);
  }

  private pageFromCrawlItem(item: Record<string, unknown>, jobId: string): FetchedPage {
    const url = (safeGet<string>(item, ["metadata", "sourceURL"]) ?? safeGet<string>(item, ["url"]) ?? "") as string;
    return new FetchedPage({
      url,
      title: safeGet<string>(item, ["metadata", "title"], "") ?? "",
      markdown: safeGet<string>(item, ["markdown"], "") ?? "",
      html: safeGet<string>(item, ["html"]) ?? null,
      metadata: { crawl_job_id: jobId, ...(safeGet<Record<string, unknown>>(item, ["metadata"], {}) ?? {}) },
      provider: this.name,
      fetchedAt: safeGet<string>(item, ["metadata", "fetchTime"]) ?? null,
      status: "ok",
    });
  }

  searchResultsFromPages(pages: FetchedPage[]): NormalizedResult[] {
    return pages
      .filter((p) => p.url)
      .map(
        (p) =>
          new NormalizedResult({
            url: p.url,
            title: p.title || p.url,
            snippet: p.markdown.slice(0, 240),
            sourceKind: "web",
            provider: this.name,
          }),
      );
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

File: cli/src/wsc/providers/firecrawl.py
"""Firecrawl — clean a URL (scrape) or crawl a site.

Endpoints (https://docs.firecrawl.dev/api-reference):

* ``POST https://api.firecrawl.dev/v1/scrape``
* ``POST https://api.firecrawl.dev/v1/crawl``  (returns a job id)
* ``GET  https://api.firecrawl.dev/v1/crawl/{id}``  (poll)

Auth: ``Authorization: Bearer <FIRECRAWL_API_KEY>``.

v0.1 implements scrape end-to-end and crawl-init only (caller drives the
poll loop via ``poll_crawl``); ``ops/crawl.py`` does a small bounded loop
guarded by ``--apply`` and the per-day budget cap.
"""

from __future__ import annotations

import time
from typing import Any, Dict, List, Optional

from wsc.providers.base import (
    FetchedPage,
    NormalizedResult,
    Provider,
    ProviderError,
    http_request,
    MissingKeyError,
    safe_get,
)

BASE_URL = "https://api.firecrawl.dev/v1"


class FirecrawlProvider(Provider):
    name = "firecrawl"
    schema_version = "firecrawl-v1-2026-04"

    def __init__(self, api_key: Optional[str] = None) -> None:
        self.api_key = api_key

    def _ensure_key(self) -> str:
        if not self.api_key:
            raise MissingKeyError(self.name)
        return self.api_key

    def _auth_headers(self) -> Dict[str, str]:
        return {"authorization": f"Bearer {self._ensure_key()}"}

    # -- scrape -----------------------------------------------------------

    def scrape(
        self,
        url: str,
        *,
        formats: Optional[List[str]] = None,
        only_main_content: bool = True,
        screenshot: bool = False,
        timeout: float = 60.0,
    ) -> FetchedPage:
        body: Dict[str, Any] = {
            "url": url,
            "formats": list(formats or ["markdown"]),
            "onlyMainContent": bool(only_main_content),
        }
        if screenshot:
            # Firecrawl returns a URL to the screenshot when this format is requested.
            if "screenshot" not in body["formats"]:
                body["formats"].append("screenshot")
        response = http_request(
            f"{BASE_URL}/scrape",
            method="POST",
            headers=self._auth_headers(),
            body=body,
            timeout=timeout,
        )
        return self._normalize_scrape(response.get("json") or {}, requested_url=url)

    def _normalize_scrape(self, payload: Dict[str, Any], *, requested_url: str) -> FetchedPage:
        if not safe_get(payload, "success", default=False):
            err = safe_get(payload, "error") or "scrape failed"
            raise ProviderError(f"firecrawl: {err}")
        data = safe_get(payload, "data", default={}) or {}
        url = safe_get(data, "metadata", "sourceURL", default=requested_url) or requested_url
        return FetchedPage(
            url=url,
            title=safe_get(data, "metadata", "title", default="") or "",
            markdown=safe_get(data, "markdown", default="") or "",
            html=safe_get(data, "html"),
            metadata=safe_get(data, "metadata", default={}) or {},
            provider=self.name,
            fetched_at=safe_get(data, "metadata", "fetchTime"),
            status="ok",
        )

    # -- crawl ------------------------------------------------------------

    def start_crawl(
        self,
        url: str,
        *,
        limit: int = 10,
        include_paths: Optional[List[str]] = None,
        exclude_paths: Optional[List[str]] = None,
        formats: Optional[List[str]] = None,
        timeout: float = 60.0,
    ) -> str:
        body: Dict[str, Any] = {
            "url": url,
            "limit": int(limit),
            "scrapeOptions": {"formats": list(formats or ["markdown"])},
        }
        if include_paths:
            body["includePaths"] = include_paths
        if exclude_paths:
            body["excludePaths"] = exclude_paths
        response = http_request(
            f"{BASE_URL}/crawl",
            method="POST",
            headers=self._auth_headers(),
            body=body,
            timeout=timeout,
        )
        payload = response.get("json") or {}
        if not safe_get(payload, "success", default=False):
            raise ProviderError(f"firecrawl: crawl failed to start: {safe_get(payload, 'error')!r}")
        job_id = safe_get(payload, "id")
        if not job_id:
            raise ProviderError(f"firecrawl: crawl response missing id: keys={list(payload.keys())}")
        return job_id

    def poll_crawl(self, job_id: str, *, timeout: float = 30.0) -> Dict[str, Any]:
        response = http_request(
            f"{BASE_URL}/crawl/{job_id}",
            method="GET",
            headers=self._auth_headers(),
            timeout=timeout,
        )
        return response.get("json") or {}

    def crawl(
        self,
        url: str,
        *,
        limit: int = 10,
        include_paths: Optional[List[str]] = None,
        exclude_paths: Optional[List[str]] = None,
        formats: Optional[List[str]] = None,
        poll_interval: float = 2.0,
        max_wait: float = 300.0,
    ) -> List[FetchedPage]:
        job_id = self.start_crawl(
            url,
            limit=limit,
            include_paths=include_paths,
            exclude_paths=exclude_paths,
            formats=formats,
        )
        deadline = time.time() + max_wait
        while time.time() < deadline:
            status = self.poll_crawl(job_id)
            state = safe_get(status, "status")
            if state in ("completed", "failed"):
                if state == "failed":
                    raise ProviderError(f"firecrawl: crawl {job_id} failed: {safe_get(status, 'error')!r}")
                pages_raw = safe_get(status, "data", default=[]) or []
                return [self._page_from_crawl_item(p, job_id=job_id) for p in pages_raw]
            time.sleep(poll_interval)
        raise ProviderError(f"firecrawl: crawl {job_id} did not complete within {max_wait}s")

    def _page_from_crawl_item(self, item: Dict[str, Any], *, job_id: str) -> FetchedPage:
        url = safe_get(item, "metadata", "sourceURL", default="") or safe_get(item, "url", default="")
        return FetchedPage(
            url=url or "",
            title=safe_get(item, "metadata", "title", default="") or "",
            markdown=safe_get(item, "markdown", default="") or "",
            html=safe_get(item, "html"),
            metadata={"crawl_job_id": job_id, **(safe_get(item, "metadata", default={}) or {})},
            provider=self.name,
            fetched_at=safe_get(item, "metadata", "fetchTime"),
            status="ok",
        )

    # -- discovery via crawl results (used by ops/discover.py for fallback) ---

    def search_results_from_pages(self, pages: List[FetchedPage]) -> List[NormalizedResult]:
        out: List[NormalizedResult] = []
        for p in pages:
            if not p.url:
                continue
            out.append(
                NormalizedResult(
                    url=p.url,
                    title=p.title or p.url,
                    snippet=(p.markdown[:240] if p.markdown else ""),
                    source_kind="web",
                    provider=self.name,
                )
            )
        return out
